StorageUtility/Linux/StorageUtility.cpp

```diff
@@ -15,253 +15,185 @@ See the License for the specific language governing permissions and
-            return( eErrorCode::Memory );
-        }
-        pDevDetailData->cbSize = sizeof( SP_INTERFACE_DEVICE_DETAIL_DATA );
+    MappingSystemPathAndDeviceHandle->s_MappingSystemPathToProperties.clear();
+    udev_list_entry_foreach(dev_list_entry, Devices)
+    {
+        const char *DevicePath;
+
+        /* Get the filename of the /sys entry for the device
+           and create a udev_device object (dev) representing it */
+        DevicePath    = udev_list_entry_get_name(dev_list_entry);
+        UdevDevice     = udev_device_new_from_syspath(UdevObject, DevicePath);
+        String devNode = udev_device_get_devnode(UdevDevice);
+        sStorageAdapterProperty property;
 
-        if ( SetupDiGetDeviceInterfaceDetail( devInfoHandle, &devInterfaceData, pDevDetailData, bufferSize, NULL, NULL ) == FALSE )
+        if( true == InitStorageAdapterProperty(property, UdevDevice) )
         {
-            error = GetLastError();
-            if ( ERROR_INSUFFICIENT_BUFFER != error )
-            {
-                HeapFree(GetProcessHeap(), HEAP_NO_SERIALIZE, pDevDetailData);
-                if ( eOnErrorBehavior::Continue == OnErrorBehavior ) { continue; }
-                SetupDiDestroyDeviceInfoList( devInfoHandle );
-                return( eErrorCode::Unknown );
-            }
+            MappingSystemPathAndDeviceHandle->s_MappingSystemPathToProperties[devNode] = property;
+            DevicePaths.push_back( devNode );
         }
 
-        eErrorCode errorCode;
-        errorCode = eErrorCode::Unknown;
-        Callback.Function(Callback.Data, devInfoHandle, devInfoData, devInterfaceData, pDevDetailData, bufferSize, errorCode);
-
-        HeapFree(GetProcessHeap(), HEAP_NO_SERIALIZE, pDevDetailData);
+        udev_device_unref(UdevDevice);
     }
+    /* Free the enumerator object */
+    udev_enumerate_unref(Enumerate);
 
-    SetupDiDestroyDeviceInfoList( devInfoHandle );
-
-    return( eErrorCode::None );
-}*/
+    udev_unref(UdevObject);
+    return( eErrorCode::None);
+}
 
-eErrorCode GetStorageAdapterProperty( HANDLE Handle, sStorageAdapterProperty& AdapterProperty )
+eErrorCode GetStorageDeviceHandle( const String& DevicePath, DeviceHandle& Handle )
 {
-   /* STORAGE_DESCRIPTOR_HEADER storageDescHeader;
-
-    STORAGE_PROPERTY_QUERY StorageProperty;
-    StorageProperty.PropertyId = StorageAdapterProperty;
-    StorageProperty.QueryType = PropertyStandardQuery;
-
-    DWORD bytesReturned;
-    DWORD ret =
-        DeviceIoControl(
-                        Handle,
-                        IOCTL_STORAGE_QUERY_PROPERTY,
-                        &StorageProperty,
-                        sizeof(STORAGE_PROPERTY_QUERY),
-                        &storageDescHeader,
-                        sizeof(STORAGE_DESCRIPTOR_HEADER),
-                        &bytesReturned,
-                        NULL
-                        );
-    if ( 0 == ret )
+    if ((Handle = open(DevicePath.c_str(), O_RDONLY|O_NONBLOCK)) < 0)
```

Comment:
Should use coding convention like this:
Handle = open(DevicePath.c_str(), O_RDONLY|O_NONBLOCK);
If ( Handle < 0) 
...